extensions/git/src/repository.ts

```diff
@@ -855,12 +855,23 @@ export class Repository implements Disposable {
 		globalState: Memento,
 		outputChannel: OutputChannel
 	) {
-		const workspaceWatcher = workspace.createFileSystemWatcher('**');
+		const workspaceWatcher = workspace.createFileSystemWatcher(new RelativePattern(repository.root, '**'));
 		this.disposables.push(workspaceWatcher);
 
-		const onWorkspaceFileChange = anyEvent(workspaceWatcher.onDidChange, workspaceWatcher.onDidCreate, workspaceWatcher.onDidDelete);
-		const onWorkspaceRepositoryFileChange = filterEvent(onWorkspaceFileChange, uri => isDescendant(repository.root, uri.fsPath));
-		const onWorkspaceWorkingTreeFileChange = filterEvent(onWorkspaceRepositoryFileChange, uri => !/\/\.git($|\/)/.test(uri.path));
+		const onWorkspaceRepositoryFileChange = anyEvent(workspaceWatcher.onDidChange, workspaceWatcher.onDidCreate, workspaceWatcher.onDidDelete);
+		const onWorkspaceWorkingTreeFileChange = filterEvent(onWorkspaceRepositoryFileChange, uri => {
+			if (/\/\.git$/.test(uri.path)) {
+				return false;
+			}
+
+			// Don't bother doing extra work, if /.git isn't in the path
+			if (!/\/\.git\//.test(uri.path)) {
+				return true;
+			}
+
+			// Normalize the uri to be relative to the repository
+			return !/(?:\/|\\)\.git(?:\/|\\)/.test(path.relative(repository.root, uri.fsPath));
```

Comment:
☝️ Possible casing issues cc @lszomoru

Reply:
Great catch! Yeah, instead of using `path.relative` can you please use the `relativePath()` helper? Thanks!

Reply:
@lszomoru What `relativePath()` helper? I see `path.relative` used in many places in `repository.ts`
@joaomoreno What casing issues? Just the drive letter normalization by `fsPath`?
@bpasero I don't think so -- if we had custom exclude rules, then yes, but I don't see how otherwise.

Reply:
I pushed a change to use a `Uri` for the `repository.root`, so now I'm using `.fsPath` for both sides of the `path.relative` so they should be more normalized.

Reply:
> What casing issues? Just the drive letter normalization

`path.relative` will return an ugly `../../` path as soon as the casing of `repository.root` is different from `uri.fsPath`. I think `repository.root` is the casing as it is on disk (returned from `git`) but `uri.fsPath` can be the casing as the user decided when opening a folder. I.e. a user can do `code <folder upper case>` and we try to preserve that even when the folder path is lowercase.

Reply:
Btw I have stopped using `path.relative` because it is somewhat overkill when you already know that a root path is equal or parent of a target path. See this change in `glob.ts`:

https://github.com/microsoft/vscode/commit/ca8d9d1542431fa4ff8271accefc58abf9a537e6#diff-95cee6e5378765ad3e4b6b91786a273b3a44b6f2ac2e19cd4ec1c9ac2094fb99R351-R359

Reply:
@bpasero so the watcher preserves the casing of how the user opened the workspace?

Reply:
@eamodio yes.

Reply:
I pushed a change to also document that. These questions are really good and help improve the documentation, even though it is getting quite long now:

https://github.com/microsoft/vscode/blob/c74b5461a9a4fab4b2aad7fb94fe3d6a9c2f54d9/src/vscode-dts/vscode.d.ts#L10999-L11008

Reply:
Pushed a change to avoid `path.relative`